src/honey_badger.rs

```diff
@@ -358,64 +350,44 @@ where
             batch.contributions.keys().collect::<Vec<_>>()
         );
         // Queue the output and advance the epoch.
-        self.output.push(batch);
-        fault_log.extend(self.update_epoch()?);
-        Ok(fault_log)
+        step.output.push_back(batch);
+        step.extend(self.update_epoch()?);
+        Ok(Some(step))
     }
 
     /// Increments the epoch number and clears any state that is local to the finished epoch.
-    fn update_epoch(&mut self) -> Result<FaultLog<NodeUid>> {
+    fn update_epoch(&mut self) -> Result<Step<C, NodeUid>> {
         // Clear the state of the old epoch.
         self.ciphertexts.remove(&self.epoch);
-        self.decrypted_contributions.clear();
         self.received_shares.remove(&self.epoch);
         self.epoch += 1;
         self.has_input = false;
         let max_epoch = self.epoch + self.max_future_epochs;
-        let mut fault_log = FaultLog::new();
+        let mut step = Step::default();
         // TODO: Once stable, use `Iterator::flatten`.
         for (sender_id, content) in
             Itertools::flatten(self.incoming_queue.remove(&max_epoch).into_iter())
         {
-            self.handle_message_content(&sender_id, max_epoch, content)?
-                .merge_into(&mut fault_log);
+            step.extend(self.handle_message_content(&sender_id, max_epoch, content)?);
         }
         // Handle any decryption shares received for the new epoch.
-        self.try_decrypt_and_output_batch()?
-            .merge_into(&mut fault_log);
-        Ok(fault_log)
+        step.extend(self.try_decrypt_and_output_batch()?);
+        Ok(step)
     }
 
     /// Tries to decrypt contributions from all proposers and output those in a batch.
-    fn try_decrypt_and_output_batch(&mut self) -> Result<FaultLog<NodeUid>> {
-        // Return if we don't have ciphertexts yet.
-        let proposer_ids: Vec<_> = match self.ciphertexts.get(&self.epoch) {
-            Some(cts) => cts.keys().cloned().collect(),
-            None => {
-                return Ok(FaultLog::new());
-            }
-        };
-
-        // Try to output a batch if all contributions have been decrypted.
-        for proposer_id in proposer_ids {
-            self.try_decrypt_proposer_contribution(proposer_id);
-        }
-        self.try_output_batch()
-    }
-
-    /// Returns true if and only if contributions have been decrypted for all selected proposers in
-    /// this epoch.
-    fn all_contributions_decrypted(&mut self) -> bool {
-        match self.ciphertexts.get(&self.epoch) {
-            None => false, // No ciphertexts yet.
-            Some(ciphertexts) => ciphertexts.keys().eq(self.decrypted_contributions.keys()),
+    fn try_decrypt_and_output_batch(&mut self) -> Result<Step<C, NodeUid>> {
```

Comment:
This function has changed quite a bit. Since `try_output_batch` is doing decryption now, it may seem strange that only this function has `decrypt` in its name. Because it is applying `try_output_batch` iteratively, maybe you could rename it to something with `iterate` or `complete` in the name instead of `decrypt`?

Reply:
Or just `try_output_batches`? `try_output_all_batches`?